test-data/unit/check-generics.test

```diff
@@ -1814,3 +1814,98 @@ def g(x: T) -> T: return x
+class A(Generic[T]):
+    @classmethod
+    def f(cls) -> T: ...
+class B(A[str]): pass
+class C(A[int]): pass
+reveal_type(B.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.int*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassMultipleInheritance]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(A[str]): pass
+class C(A[int]): pass
+reveal_type(B.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(B.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.int*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.int*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassMultipleLevelInheritance]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(A[str]): pass
+class C(B): pass
+class D(C): pass
+class E(D): pass
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(D.f())  # E: Revealed type is 'builtins.str*'
+reveal_type(E.f)  # E: Revealed type is 'def () -> builtins.str*'
+reveal_type(E.f())  # E: Revealed type is 'builtins.str*'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfMultipleGenericClass]
+from typing import TypeVar, Generic
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class B(Generic[t]):
+    @classmethod
+    def f(cls) -> t: ...
+class C(A[str], B[int]): pass
+class D(A[int], B[str]): pass
+reveal_type(C.f)  # E: Revealed type is 'def () -> builtins.str'
+reveal_type(C.f())  # E: Revealed type is 'builtins.str'
+reveal_type(D.f)  # E: Revealed type is 'def () -> builtins.int'
+reveal_type(D.f())  # E: Revealed type is 'builtins.int'
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassOverriding]
+from typing import Generic, TypeVar
+t = TypeVar('t')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: pass
+class B(A[str]):
+    @classmethod
+    def f(cls) -> str: pass
+class C(A[str]):
+    @classmethod
+    def f(cls) -> int: pass  # E: Return type of "f" incompatible with supertype "A"
+[builtins fixtures/classmethod.pyi]
+[out]
+
+[case testClassMethodOfGenericClassWithBindingAnotherTypeVariable]
+from typing import Generic, TypeVar
+t = TypeVar('t')
+s = TypeVar('s')
+class A(Generic[t]):
+    @classmethod
+    def f(cls) -> t: pass
+class B(A[s]): pass
+reveal_type(A.f())  # E: Revealed type is '<nothing>'
+reveal_type(B.f())  # E: Revealed type is '<nothing>'
+[builtins fixtures/classmethod.pyi]
+[out]
```

Comment:
Here suggestions for more tests:
* Overloaded class methods
* More complex bindings, like `T -> Iterable[T] -> Iterable[str]`, etc.
* More complex return types (like `Iterable[Tuple[T, T]]` etc.) and argument types
* Basic multiple inheritance
* Overriding between bases in multiple inheritance

In addition, as I mentioned above:
* Give an error if class method accessed via _class object_ contains type variables after everything is bound. Similar to what we do for variables
* Add tests for binding types of _variables_ and normal methods accessed on class object.